### ISSUE TITLE
Fix false negatives for object notation and style-search rules

### ISSUE THREAD

**Issue by swernerx** [stylelint/stylelint#4826]
> Clearly describe the bug

When using the jss-react syntax I see that only some rules are working properly. Interestingly this seems to be specific to object notation in JSS. I tried debugging this deep into the code but was not very successful even after deep inside `style-search`

> Which rule, if any, is the bug related to?

The most prominent rules we figured out are not working correctly are:

```
length-zero-no-unit: true
color-hex-length: short
```

> What code is needed to reproduce the bug?

```js
import { createUseStyles } from "react-jss"

const useStyles = createUseStyles({
  button: {
    borderRadius: '0px',
    font: "Arial",
    fontSize: 12,
    background: '#2200DD'
  }
})
```

> What stylelint configuration is needed to reproduce the bug?

```yaml
rules:
  length-zero-no-unit: true
  color-hex-length: short
  color-hex-case: lower
  unit-case: lower
```

> Which version of stylelint are you using?

`13.6.0`

> How are you running stylelint: CLI, PostCSS plugin, Node.js API?

CLI, just the basic `stylelint --fix filename.js`

> Does the bug relate to non-standard syntax (e.g. SCSS, Less etc.)?

It is related to CSS-in-JS, in this case a JSS based syntax.

> What did you expect to happen?

Reporting the same issue like when the code was written in a template literal.

```js
import styled from "styled-components"

const Button = styled.button`
  border-radius: 0px;
  font: Arial;
  font-size: 12;
  background: #2200DD;
`
```

results in this output:

```shell
 4:19  ✖  Unexpected unit                          length-zero-no-unit
 5:9   ✖  Unexpected missing generic font family   font-family-no-missing-generic-family-keyword
 7:15  ✖  Expected "#2200DD" to be "#2200dd"       color-hex-case
 7:15  ✖  Expected "#2200DD" to be "#20D"          color-hex-length
```

> What actually happened (e.g. what warnings or errors did you get)?

Interestingly the font-family warning came up. So it seems to work at least in some part. But all of `color-hex-length`, `color-hex-case` and `unit-case` are not detected correctly.

```shell
npx stylelint --fix filename.js
6:11  ✖  Unexpected missing generic font family   font-family-no-missing-generic-family-keyword
```

**Comment by swernerx** [comment]
There is a difference in how `styleSearch` is called:

- JSS: source=`borderRadius: '0px'`; target=`0`
- Styled Components: source=`border-radius: 0px`; target=`0`

The quotes might be the root cause of that issue.

**Comment by swernerx** [comment]
Nodes in JSS:

```json
  nodes: [
    { type: 'word', sourceIndex: 0, value: 'borderRadius' },
    {
      type: 'div',
      sourceIndex: 12,
      value: ':',
      before: '',
      after: ' '
    },
    { type: 'string', sourceIndex: 14, quote: "'", value: '0px' }
  ]
```

Same in Styled Components:

```
  nodes: [
    { type: 'word', sourceIndex: 0, value: 'font' },
    { type: 'div', sourceIndex: 4, value: ':', before: '', after: ' ' },
    { type: 'word', sourceIndex: 6, value: 'Arial' }
  ]
```

**Comment by jeddy3** [comment]
@swernerx Thank you for the clear report, using the template and digging into the issue.

I think you're right to identify`style-search` as a possible culprit here. The rules in question are some of the earliest ones we wrote, and `style-search` was written to make them possible. Shortly after, a [value parser was written for PostCSS](https://www.npmjs.com/package/postcss-value-parser) and most of the newer rules use that. It seems more reliable when working with non-CSS-like syntax.

Rather than delving deeper into `style-search`, I think we should update the rules to use the value parser. Fortunately, all the rules are accompanied by comprehensive test suites, so I believe we can update with confidence. What is missing from the test suites are one or two JSS test cases, which we should add while updating the rules.

(I'm unable to reproduce the issue for `unit-case` [using the demo](https://stylelint.io/demo#N4Igxg9gJgpiBcICWBbADhATgFwATFzExgENsYBVAZxgGVsBPAGxitwF9cAzTCFXADohiJMNgC0AKypUhAgHYLI8qngCuNes1a4AvIRHlqdRiyoAKYAty4ARmuzYI8+Pms27WWJgBKJKEgargDkAAwACgAawQA07jZcztiuQgCCmEgkTEJx8h7cSbRIAF4wrgCMAEy5+baiANYA5rxq8lAhAMSVlaGhACJ9sfGEEExYIQCSAHIAEgCiPhMAKsHu7ArsAJQgMSBUDPLYJAAeCOAy4kjyUlQ74M5cSI1nVnmCwmpmQq6v+UKQY0w4gAFjBjuIwCQaN93mMAO4wTA5Yb-UZYEFg8QseSNbDAmFCKjArDYZFvGxCbG44HiUq8cTyCDiVpIUkIXDYTBqGAxQTk94siSQ6HsykQBFInYokAANyy3PE9RgDDhXghUJgBJA8MRcje63k7BA7CAA). As this rule uses the value parser already, I think this validates the suggestion to update the other rules.)

I'll label this up as an umbrella issue for:

- [ ] `length-zero-no-unit`
- [ ] `color-hex-length`
- [ ] `color-hex-case`

I've labelled the issue as ready to implement. Please consider [contributing](https://stylelint.io/contributing) if you have time.

There are [steps on how to fix a bug in a rule](https://stylelint.io/developer-guide/rules#fix-a-bug-in-a-rule) in the Developer guide.

**Comment by Dru89** [comment]
This still seems to be a problem in the latest version.

I'd love to take a shot at replacing some of the instances of `style-search` usage with the postcss value parser. Is that still what you'd recommend doing here?

**Comment by jeddy3** [comment]
@Dru89 Thanks for wanting to take a shot. Yes, replacing `style-search` with the value parser is still what's recommended. 

The [`color-hex-case` rule](https://github.com/stylelint/stylelint/blob/master/lib/rules/color-hex-case/index.js) would be a good one to start with. 

The [`color-function-notation`](https://github.com/stylelint/stylelint/blob/master/lib/rules/color-function-notation/index.js) and [`alpha-value-notation`](https://github.com/stylelint/stylelint/blob/master/lib/rules/alpha-value-notation/index.js) rules were recently added, and they use the value parser and support autofix. I suggest looking at them for tips on working with the parser, e.g. make use of the `setValue` and `getValue` functions otherwise comments will be stripped.

**Comment by Dru89** [comment]
Thanks for the tips, @jeddy3! That's along the lines of the first rule I was gonna take a look at. I'll try to get something up this week, and I'll add some tests to `color-hex-case` that use `css-in-js` and object notation to make sure this explicit problem is also fixed.

**Comment by jeddy3** [comment]
Closing as the [`syntax` option was removed](https://github.com/stylelint/stylelint/issues/5289) for the [14.0.0 release](https://github.com/stylelint/stylelint/issues/5205). (See the [migration guide](https://github.com/stylelint/stylelint/blob/main/docs/migration-guide/to-14.md#syntax-option-and-automatic-inferral-of-syntax).)

We'll also [be deprecating](https://github.com/stylelint/postcss-css-in-js/issues/225) our [forked CSS-in-JS package](https://github.com/stylelint/postcss-css-in-js), in favour of leaner custom syntaxes.

Please consider [writing a custom syntax](https://github.com/stylelint/stylelint/blob/main/docs/developer-guide/syntaxes.md) specific to the CSS-in-JS library in this issue and fixing the bug there.
